engine/common/message.h

```diff
@@ -0,0 +1,204 @@
+#pragma once
+
+#include <json/reader.h>
+#include <json/value.h>
+#include <json/writer.h>
+#include <cstdint>
+#include <string>
+#include "common/message_attachment.h"
+#include "common/message_attachment_factory.h"
+#include "common/message_content.h"
+#include "common/message_content_factory.h"
+#include "common/message_incomplete_detail.h"
+#include "common/message_role.h"
+#include "common/message_status.h"
+#include "common/variant_map.h"
+#include "json_serializable.h"
+#include "utils/logging_utils.h"
+#include "utils/result.hpp"
+
+namespace ThreadMessage {
+
+// Represents a message within a thread.
+struct Message : JsonSerializable {
+
+  // The identifier, which can be referenced in API endpoints.
+  std::string id;
+
+  // The object type, which is always thread.message.
+  std::string object = "thread.message";
+
+  // The Unix timestamp (in seconds) for when the message was created.
+  uint32_t created_at;
+
+  // The thread ID that this message belongs to.
+  std::string thread_id;
+
+  // The status of the message, which can be either in_progress, incomplete, or completed.
+  Status status;
+
+  // On an incomplete message, details about why the message is incomplete.
+  std::optional<IncompleteDetail> incomplete_details;
+
+  // The Unix timestamp (in seconds) for when the message was completed.
+  std::optional<uint32_t> completed_at;
+
+  // The Unix timestamp (in seconds) for when the message was marked as incomplete.
+  std::optional<uint32_t> incomplete_at;
+
+  Role role;
+
+  // The content of the message in array of text and/or images.
+  std::vector<std::unique_ptr<Content>> content;
```

Comment:
Not sure why do we need `unique_ptr`?

Reply:
because `Content` is a pure virtual class. c++ STL containers are not able to handle polymorphic